x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -941,72 +1005,49 @@ public void onFailure(Exception e) {
-                        listener.onFailure(new ElasticsearchSecurityException("could not refresh the requested token", e));
-                        return;
-                    }
-                    listener.onResponse(parsedTokens);
-                } else {
-                    // We retry this since the creation of the superseding token document might already be in flight but not
-                    // yet completed, triggered by a refresh request that came a few milliseconds ago
-                    logger.info("could not find superseding token document from [{}] reference, retrying",
-                            refreshTokenStatus.getSupersededBy());
-                    maybeRetryOnFailure.accept(invalidGrantException("could not refresh the requested token"));
-                }
-            }
-
-            @Override
-            public void onFailure(Exception e) {
-                if (isShardNotAvailableException(e)) {
-                    logger.info("could not find superseding token document from reference [{}], retrying",
-                            refreshTokenStatus.getSupersededBy());
-                    maybeRetryOnFailure.accept(invalidGrantException("could not refresh the requested token"));
-                } else {
-                    logger.warn("could not find superseding token document from reference [{}]", refreshTokenStatus.getSupersededBy());
-                    onFailure.accept(invalidGrantException("could not refresh the requested token"));
-                }
-            }
-        });
+    /**
+     * Decrypts the values of the superseding access token and the refresh token, using a key derived from the superseded refresh token. It
+     * encodes the version and serializes the tokens before calling the listener, in the same manner as {@link #createOAuth2Tokens } does
+     **/
+    void decryptAndReturnSupersedingTokens(String refreshToken, RefreshTokenStatus refreshTokenStatus,
+                                           ActionListener<Tuple<String, String>> listener) {
+        final byte[] iv = Base64.getDecoder().decode(refreshTokenStatus.getIv());
+        final byte[] salt = Base64.getDecoder().decode(refreshTokenStatus.getSalt());
+        final byte[] encryptedSupersedingAccessToken = Base64.getDecoder().decode(refreshTokenStatus.getSupersedingAccessToken());
+        final byte[] encryptedSupersedingRefreshToken = Base64.getDecoder().decode(refreshTokenStatus.getSupersedingRefreshToken());
+        try {
+            Cipher cipher = getDecryptionCipher(iv, refreshToken, salt);
+            final String supersedingAccessToken = new String(cipher.doFinal(encryptedSupersedingAccessToken), StandardCharsets.UTF_8);
+            byte[] iv2 = new byte[iv.length];
+            System.arraycopy(iv, iv.length / 2, iv2, 0, iv.length / 2);
+            System.arraycopy(iv, 0, iv2, iv.length / 2, iv.length / 2);
+            cipher = getDecryptionCipher(iv2, refreshToken, salt);
```

Comment:
Per your most recent comment on the PR, I think it definitely makes sense to do a single encrypt of the concatenated tokens.